### PR TITLE
Noncollinear and spin-orbit implementation for k-point Hamiltonian in AFQMC

### PR DESCRIPTION
**Pull request by mmorale3**
Please review the [developer documentation](https://github.com/QMCPACK/qmcpack/wiki/Development-workflow)
on the wiki of this project that contains help and requirements.

The non-collinear implementation of AFQMC is completed in this PR.
In non-collinear calculations, a spin-orbit hamiltonian is now allowed.
Propagation and energy evaluation are implemented.
Only the k-point hamiltonian (both CPU and GPU) works with non-collinear.
Observables (other than the energy) are not yet enabled with non-collinear.
Future PRs will complete the missing features.

Describe what this PR changes and why.  If it closes an issue, link to it here
with [a supported keyword](https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword).

New feature: Noncollinear and spin-orbit interactions in AFQMC.

### Does this introduce a breaking change?

No

## What systems has this change been tested on?

Quartz, Lassen. GCC and Clang compilers.

## Checklist

- Yes This PR is up to date with current the current state of 'develop'

### CODE REVIEW

File: external_codes/boost_multi/multi/memory/allocator.hpp [hunk 1]
```diff
@@ -49,7 +49,8 @@ class allocator{
 	bool operator!=(allocator const& o) const{return mp_ != o.mp_;}
 	using void_pointer = typename std::pointer_traits<decltype(std::declval<memory_type*>()->allocate(0, 0))>::template rebind<void>;
 	pointer allocate(size_type n){
-		return static_cast<pointer>(static_cast<void_pointer>(mp_->allocate(n*sizeof(value_type), alignof(T))));
+		return static_cast<pointer>(static_cast<void_pointer>(mp_->allocate(n*sizeof(value_type), 16)));
```
Comment (fdmalone) [inline external_codes/boost_multi/multi/memory/allocator.hpp:52]: why 16?

**Review by fdmalone — APPROVED** [review]
LGTM

**Comment by fdmalone** [comment]
This only allows for spin dependent one-body Hamiltonian for k-point calculations right? What format  is expected in the input wavefunction / hamiltonian.

File: external_codes/boost_multi/multi/memory/allocator.hpp (hunk 1 above)
Comment (mmorale3) [reply external_codes/boost_multi/multi/memory/allocator.hpp:52]: This is a temporary fix until we update multi.
Cuda is incompatible with alignof() for std::complex<double>, so using alignof(T) in general
causes alignment issues when compiled in double precision for gpus.
With 16 all the regular types work fine. (Notice that cudamalloc aligns to 256).

**Comment by mmorale3** [comment]
The one body hamiltonian as to be 2M*2M in non-collinear calculations.
Only the kpoint hamiltonian works now. For any other hamiltonian, the code should abort at some point
when it realizes the calculation is non-collinear. I will complete the missing parts soon.

**Comment by fdmalone** [comment]
For the kpoint code this is 2*m x 2*m right? The HDF5 format doesn't change?

File: external_codes/boost_multi/multi/memory/allocator.hpp (hunk 1 above)
Comment (ye-luo) [reply external_codes/boost_multi/multi/memory/allocator.hpp:52]: Should we update multi?

**Review by ye-luo — CHANGES_REQUESTED** [review]

**Comment by mmorale3** [comment]
Updating multi is possibly a lot of work.
I prefer to push this first and update multi after.

**Review by ye-luo — APPROVED** [review]
Preferably, we should update multi and then update our code. I tried updating multi but found an internal conflict in multi. So let us use the workaround from @mmorale3 at the moment.

**Pull request merged**